File: problems/0023-merge-k-sorted-lists/solution.js
/**
 * @problemNumber: 23
 * @problemName: Merge k Sorted Lists
 * @difficulty: Hard
 * @topics: Linked List, Divide and Conquer, Heap
 * @companies: Facebook, Amazon, Google
 */

function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
}

/**
 * Approach: Divide and Conquer
 * Time Complexity: O(N log k)
 * Space Complexity: O(1)
 *
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
function mergeKLists(lists) {
    if (!lists || lists.length === 0) return null;

    while (lists.length > 1) {
        const mergedLists = [];
        for (let i = 0; i < lists.length; i += 2) {
            const l1 = lists[i];
            const l2 = i + 1 < lists.length ? lists[i + 1] : null;
            mergedLists.push(mergeTwoLists(l1, l2));
        }
        lists = mergedLists;
    }

    return lists[0];
}

function mergeTwoLists(l1, l2) {
    const dummy = new ListNode(0);
    let tail = dummy;

    while (l1 && l2) {
        if (l1.val <= l2.val) {
            tail.next = l1;
            l1 = l1.next;
        } else {
            tail.next = l2;
            l2 = l2.next;
        }
        tail = tail.next;
    }

    if (l1) tail.next = l1;
    if (l2) tail.next = l2;

    return dummy.next;
}

// Helper to create list
function createList(arr) {
    let dummy = new ListNode(0);
    let curr = dummy;
    for (let val of arr) {
        curr.next = new ListNode(val);
        curr = curr.next;
    }
    return dummy.next;
}

// Helper to array
function toArray(head) {
    let res = [];
    while (head) {
        res.push(head.val);
        head = head.next;
    }
    return res;
}

// Test cases
let l1 = createList([1, 4, 5]);
let l2 = createList([1, 3, 4]);
let l3 = createList([2, 6]);
console.log(toArray(mergeKLists([l1, l2, l3]))); // [1,1,2,3,4,4,5,6]

module.exports = { mergeKLists };

File: problems/0023-merge-k-sorted-lists/solution.py
"""
Problem Number: 23
Problem Name: Merge k Sorted Lists
Difficulty: Hard
Topics: Linked List, Divide and Conquer, Heap
Companies: Facebook, Amazon, Google
"""

from typing import List, Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        """
        Approach: Divide and Conquer
        Time Complexity: O(N log k)
        Space Complexity: O(1)
        """
        if not lists or len(lists) == 0:
            return None
            
        while len(lists) > 1:
            merged_lists = []
            for i in range(0, len(lists), 2):
                l1 = lists[i]
                l2 = lists[i+1] if i + 1 < len(lists) else None
                merged_lists.append(self.mergeTwoLists(l1, l2))
            lists = merged_lists
            
        return lists[0]

    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)
        tail = dummy
        
        while list1 and list2:
            if list1.val <= list2.val:
                tail.next = list1
                list1 = list1.next
            else:
                tail.next = list2
                list2 = list2.next
            tail = tail.next
            
        if list1:
            tail.next = list1
        elif list2:
            tail.next = list2
            
        return dummy.next

# Helper methods for testing
def create_list(arr):
    dummy = ListNode(0)
    curr = dummy
    for val in arr:
        curr.next = ListNode(val)
        curr = curr.next
    return dummy.next

def list_to_arr(head):
    res = []
    while head:
        res.append(head.val)
        head = head.next
    return res

# Test cases
if __name__ == "__main__":
    solution = Solution()
    l1 = create_list([1,4,5])
    l2 = create_list([1,3,4])
    l3 = create_list([2,6])
    res = solution.mergeKLists([l1, l2, l3])
    print(list_to_arr(res)) # [1,1,2,3,4,4,5,6]
